text/3808-register-tool.md

```diff
@@ -0,0 +1,257 @@
+
+In the context of an attribute, ambiguity between a tool name and a local item is always a hard error. For example, this code would error:
+
+```rust
+#![register_tool(name)]
+
+mod inner {
+    mod name {
+        // Import the derive macro.
+        use Clone as x;
+    }
+
+    #[name::x]        // ERROR: ambiguous
+    #[name::y]        // ERROR: ambiguous (even though y is not present)
+    #[self::name::x] // OK
+    fn f() {}
+}
+```
+This is in order to not require external tools to perform name resolution. This restriction may be relaxed in the future to favor tool names.
+
+The `#![register_attribute_tool(ident)]` crate-level attribute adds a new tool to the tool attribute prelude. The `#![register_lint_tool(ident)]` crate-level attribute adds a new tool to the lint prelude. For both attributes, tools must be a single ident, not a nested path.
+
+`#[no_implicit_prelude]` keeps its current behavior, i.e. it removes all tools that are in the prelude by default. However, those tools can now be explicitly added back with e.g. `#![register_attribute_tool(rustfmt)]`.
+
+Like today, attributes and lints in a tool namespace are always considered used by the compiler. The compiler does not verify the contents of any tool attribute, except to verify that all attributes are syntactically valid [tool attributes].
+
+Unknown tool names in lints remain a hard error until the story for proc-macro lints is resolved (see [Future possibilities](#future-possibilities)).
+
+Modules in the first path of an attribute (e.g. `#[unregistered::name]`) are assumed to be a crate if they cannot be resolved, and therefore give a hard error if not registered.
+
+Registering a tool multiple times is an error. This makes registering a predefined tool (`clippy`, `miri`) using `#![register_*_tool(...)]` an error unless `#![no_implicit_prelude]` is specified.
+
+[tool attributes]: https://doc.rust-lang.org/nightly/reference/attributes.html#tool-attributes
+[`unknown_lints`]: https://doc.rust-lang.org/rustc/lints/listing/warn-by-default.html#unknown-lints
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+This makes the rules for name resolution even more complicated.
+
+This runs the risk that external tools will add attributes that change the meaning of the code, such that the behavior is different when the tool is present. There is not much we can do about this other than to ask tool authors not to do that.
+
+This introduces a new "meta-breaking" concern: once this is stabilized, adding a new tool namespace, like we did for `diagnostic` and `rust_analyzer`, becomes a breaking change. I think in practice there will be few enough external tools that this will not be a major concern and we can just pick a name that isn't already in use. Note that, because `![register_*_tool(...)]` for a built-in tool errors, the user will see immediately what went wrong, instead of getting confusing errors from the attributes in the tool namespace.
```

Comment:
Should we pre-emptively reserve some namespaces?

Reply:
That seems like a good idea - at least `metadata` is something I'd like to reserve. It's much easier to reserve them now than to try and add them later.

Reply:
I'd rather go the opposite way and remove all the hardcoded tools on the next edition, and always require registering them explicitly.
There are zero reasons for rustc to know about, e.g. `rust_analyzer`.

Reply:
>at least `metadata` is something I'd like to reserve

This reminded me of something - https://github.com/rust-lang/rust/pull/75802.

Reply:
oh, "adding a new namespace has to be over an edition boundary" seems very reasonable, yes. that avoids having to treat built-in tools differently from registered tools.